File: agent/crates/event-buffer/tests/event_buffer_test.rs
use event_buffer::EventBuffer;

#[tokio::test]
async fn test_event_buffer_flow() {
    let dir = tempfile::tempdir().unwrap();
    let db_path = dir.path().join("events.db");

    // Create buffer with max_events = 3
    let buffer = EventBuffer::new(&db_path, 3).unwrap();

    // Check empty
    assert!(buffer.is_empty().await.unwrap());
    assert_eq!(buffer.len().await.unwrap(), 0);

    // Push 4 events (should evict the first one because max_events=3)
    buffer.push("event 1".to_string()).await.unwrap();
    buffer.push("event 2".to_string()).await.unwrap();
    buffer.push("event 3".to_string()).await.unwrap();
    buffer.push("event 4".to_string()).await.unwrap();

    // Check length is 3 (max capacity)
    assert_eq!(buffer.len().await.unwrap(), 3);

    // Drain 2 events
    let drained = buffer.drain(2).await.unwrap();
    assert_eq!(drained.len(), 2);

    // Since event 1 was evicted, the oldest remaining is event 2
    assert_eq!(drained[0], "event 2");
    assert_eq!(drained[1], "event 3");

    // Length should now be 1
    assert_eq!(buffer.len().await.unwrap(), 1);

    // Drain remaining
    let remaining = buffer.drain(5).await.unwrap();
    assert_eq!(remaining.len(), 1);
    assert_eq!(remaining[0], "event 4");

    // Empty again
    assert!(buffer.is_empty().await.unwrap());
}

File: agent/crates/event_buffer/tests/event_buffer_test.py
import tempfile
import unittest
from pathlib import Path

from event_buffer import EventBuffer


class TestEventBuffer(unittest.IsolatedAsyncioTestCase):

    async def test_event_buffer_flow(self):

        with tempfile.TemporaryDirectory() as temp_dir:

            db_path = Path(temp_dir) / "events.db"

            # Create buffer with max_events = 3
            buffer = EventBuffer(db_path, 3)

            # Check empty
            self.assertTrue(await buffer.is_empty())
            self.assertEqual(await buffer.len(), 0)

            # Push 4 events (oldest should be evicted)
            await buffer.push("event 1")
            await buffer.push("event 2")
            await buffer.push("event 3")
            await buffer.push("event 4")

            # Capacity should be 3
            self.assertEqual(await buffer.len(), 3)

            # Drain 2 events
            drained = await buffer.drain(2)

            self.assertEqual(len(drained), 2)
            self.assertEqual(drained[0], "event 2")
            self.assertEqual(drained[1], "event 3")

            # One event should remain
            self.assertEqual(await buffer.len(), 1)

            remaining = await buffer.drain(5)

            self.assertEqual(len(remaining), 1)
            self.assertEqual(remaining[0], "event 4")

            # Buffer should now be empty
            self.assertTrue(await buffer.is_empty())
